File: DataStructures/SparseSet.js
"use strict";

class SparseSet1 {

  constructor(capacity = 10, maxVal = 10) {

    this.dense = new Array(capacity);
    this.sparse = new Array(maxVal + 1);
    this.capacity = capacity;
    this.maxVal = maxVal;
    this.items = 0;

  }

  add(item) {

    if (item > this.maxVal || this.items >= this.capacity || this.find(item) != -1) {
      return;
    }

    this.dense[this.items] = item;
    this.sparse[item] = this.items;

    this.items++;

  }

  remove(item) {

    if (this.find(item) == -1) {
      return;
    }

    let temp = this.dense[this.items - 1];
    this.dense[this.sparse[item]] = temp;
    this.sparse[temp] = this.sparse[item];

    this.items--;

  }

  find(item) {

    if (item <= this.maxVal && this.sparse[item] < this.items && this.dense[this.sparse[item]] == item) {
      return this.sparse[item];
    }

    return -1;

  }

  clear() {

    this.items = 0;

  }

}

for (let sparseSetClass of [SparseSet1]) {

  let sparseSet = new sparseSetClass();

  sparseSet.add(2);
  sparseSet.add(4);

  console.log(sparseSet.find(4));
  console.log(sparseSet.find(6));

  sparseSet.add(6);
  sparseSet.add(8);

  console.log(sparseSet.find(6));

  sparseSet.remove(6);

  console.log(sparseSet.find(6));
  console.log(sparseSet.find(8));

  sparseSet.clear();

  console.log(sparseSet.find(2));
  console.log(sparseSet.find(8));

  console.log("------");

}

File: DataStructures/SparseSet.java


public class SparseSet {

    private interface SparseSetDataStructure {

        public void add(int item);

        public void remove(int item);

        public int find(int item);

        public void clear();

    }

    private static class SparseSet1 implements SparseSetDataStructure {

        int items, capacity, maxVal;
        int[] dense, sparse;

        public SparseSet1(int cap, int maxV) {

            dense = new int[cap];
            sparse = new int[maxV + 1];
            capacity = cap;
            maxVal = maxV;
            items = 0;

        }

        public void add(int item) {

            if(item > maxVal || items >= capacity || find(item) != -1) {
                return;
            }

            dense[items] = item;
            sparse[item] = items;

            items++;

        }

        public void remove(int item) {

            if(find(item) == -1) {
                return;
            }

            int temp = dense[items - 1];
            dense[sparse[item]] = temp;
            sparse[temp] = sparse[item];

            items--;

        }

        public int find(int item) {

            if(item <= maxVal && sparse[item] < items && dense[sparse[item]] == item) {
                return sparse[item];
            }

            return -1;

        }

        public void clear() {

            items = 0;

        }

    }

    public static void main(String[] args) {

        SparseSetDataStructure[] setClasses = new SparseSetDataStructure[]{new SparseSet1(10, 10)};

        for (SparseSetDataStructure sparseSet : setClasses) {

            sparseSet.add(2);
            sparseSet.add(4);

            System.out.println(sparseSet.find(4));
            System.out.println(sparseSet.find(6));

            sparseSet.add(6);
            sparseSet.add(8);

            System.out.println(sparseSet.find(6));

            sparseSet.remove(6);

            System.out.println(sparseSet.find(6));
            System.out.println(sparseSet.find(8));

            sparseSet.clear();

            System.out.println(sparseSet.find(2));
            System.out.println(sparseSet.find(8));

            System.out.println("------");

        }

    }

}
